### PR TITLE
[#555] health monitoring example cxx

### LINKED ISSUE #555: health monitoring example
## Brief feature description

Example that demonstrates the tools that are available in iceoryx2 to monitor the health state of a system.

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

* [x] Add sensible notes for the reviewer
* [x] PR title is short, expressive and meaningful
* [x] Consider switching the PR to a draft (`Convert to draft`)
    * as draft PR, the CI will be skipped for pushes
* [x] Relevant issues are linked in the [References](#references) section
* [x] Every source code file has a copyright header with `SPDX-License-Identifier: Apache-2.0 OR MIT`
* [x] Branch follows the naming format (`iox2-123-introduce-posix-ipc-example`)
* [x] Commits messages are according to this [guideline][commit-guidelines]
    * [x] Commit messages have the issue ID (`[#123] Add posix ipc example`)
    * [x] Commit author matches [Eclipse Contributor Agreement][eca](and ECA is signed)
* [x] Tests follow the [best practice for testing][testing]
* [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
* [x] Assign PR to reviewer
* [x] All checks have passed (except `task-list-completed`)

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/doc/release-notes/iceoryx2-unreleased.md

## Checklist for the PR Reviewer

* [x] Commits are properly organized and messages are according to the guideline
* [x] Unit tests have been written for new behavior
* [x] Public API is documented
* [x] PR title describes the changes

## Post-review Checklist for the PR Author

* [x] All open points are addressed and tracked via issues

## References

Closes #555

### CODE REVIEW

**Review by elBoberido — COMMENTED** [review]
Looks good. Once the CI und s green, this can be merged

File: .github/workflows/build-test.yml [hunk 1]
```diff
@@ -333,7 +333,7 @@ jobs:
 
       - name: Build language bindings
         run: |
-          cmake -S . -B target/ffi/build -DBUILD_EXAMPLES=ON -DBUILD_TESTING=ON ${{ matrix.mode.cmake-build-type }} ${{ matrix.mode.cmake-cxx-flags }} ${{ matrix.cmake-build-system-generator }} -DCMAKE_INSTALL_PREFIX=target/ffi/install -DCMAKE_PREFIX_PATH="${{ github.workspace }}/target/iceoryx/install"
+          cmake -S . -B target/ffi/build -DBUILD_EXAMPLES=ON -DBUILD_TESTING=ON -DTREAT_WARNING_AS_ERROR=ON ${{ matrix.mode.cmake-build-type }} ${{ matrix.mode.cmake-cxx-flags }} ${{ matrix.cmake-build-system-generator }} -DCMAKE_INSTALL_PREFIX=target/ffi/install -DCMAKE_PREFIX_PATH="${{ github.workspace }}/target/iceoryx/install"
```
Comment (elBoberido) [inline .github/workflows/build-test.yml:336]: Should we stay closer to what we have in iceoryx classic? There the flag is called `BUILD_STRICT`

File: CMakeLists.txt [hunk 2]
```diff
@@ -104,6 +110,11 @@ add_rust_feature(
     RUST_FEATURE "iceoryx2/logger_tracing"
 )
 
+if(TREAT_WARNING_AS_ERROR)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Werror -Wall -Wextra -Wpedantic")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror -Wall -Wextra -Wpedantic")
+endif()
```
Comment (elBoberido) [inline CMakeLists.txt:116]: We need to handle this closer to iceoryx classic and enable those warnings by default. Not for this PR, though.

File: iceoryx2-ffi/cxx/src/port_factory_event.cpp [hunk 3]
```diff
@@ -81,6 +81,7 @@ auto PortFactoryEvent<S>::dynamic_config() const -> const DynamicConfigEvent& {
 template <ServiceType S>
 auto PortFactoryEvent<S>::nodes(const iox::function<CallbackProgression(NodeState<S>)>& callback) const
     -> iox::expected<void, NodeListFailure> {
+    static_cast<void>(callback);
```
Comment (elBoberido) [inline iceoryx2-ffi/cxx/src/port_factory_event.cpp:84]:
```suggestion
auto PortFactoryEvent<S>::nodes([[maybe_unused]] const iox::function<CallbackProgression(NodeState<S>)>& callback) const
    -> iox::expected<void, NodeListFailure> {
```

File: iceoryx2-ffi/cxx/src/service.cpp [hunk 4]
```diff
@@ -42,6 +42,9 @@ auto Service<S>::details(const ServiceName& service_name,
                          const ConfigView config,
                          const MessagingPattern messaging_pattern)
     -> iox::expected<iox::optional<ServiceDetails<S>>, ServiceDetailsError> {
+    static_cast<void>(service_name);
+    static_cast<void>(config);
+    static_cast<void>(messaging_pattern);
```
Comment (elBoberido) [inline iceoryx2-ffi/cxx/src/service.cpp:47]:
```suggestion
                         [[maybe_unused]] const ConfigView config,
                         [[maybe_unused]] const MessagingPattern messaging_pattern)
    -> iox::expected<iox::optional<ServiceDetails<S>>, ServiceDetailsError> {
```
same with the first argument

File: .github/workflows/build-test.yml (hunk 1 above)
Comment (elfenpiff) [reply .github/workflows/build-test.yml:336]: Honestly, I always have to scratch my head and have to figure out `BUILD_STRICT` really means. `TREAT_WARNING_AS_ERROR` is straight forward.

If you don't mind, I would go with it.

Comment (elBoberido) [reply .github/workflows/build-test.yml:336]: The idea was to be able to add more constraints than just the warning as error, without having to add more parameter.

What about shortening the parameter to just `WARNING_AS_ERROR`?

**Review by elBoberido — DISMISSED** [review]

**Review by elBoberido — APPROVED** [review]

**Pull request merged**